src/surface_header.F90

```diff
@@ -183,6 +183,15 @@ end function iNormal
     procedure :: normal => z_cone_normal
   end type SurfaceZCone
 
+  type, extends(Surface) :: SurfaceQuadric
+    ! Ax^2 + By^2 + Cz^2 + Dxy + Eyz + Fxz + Gx + Hy + Jz + K = 0
+    real(8) :: A, B, C, D, E, F, G, H, J, K
+  contains
+    procedure :: evaluate => quadric_evaluate
```

Comment:
The `evaluate`, `distance`, and `normal` procedures are polymorphic (like a virtual method in C++), so if you look at the corresponding code in the geometry module, you'll see that there are no `select type` blocks. For example, in distance_to_boundary, we have the following line:
```fortran
d = surf%distance(...)
```
Of course, under the hood, the type needs to be determined at run-time which does incur some overhead.